src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,770 @@
+{
+    uint32_t delay;
+
+    if (aValidLifetime * static_cast<uint64_t>(1000) > Timer::kMaxDelay)
+    {
+        delay = Timer::kMaxDelay;
+    }
+    else
+    {
+        delay = aValidLifetime * 1000;
+    }
+
+    return delay;
+}
+
+void RoutingManager::HandleRouterAdvertisement(const Ip6::Address &aSrcAddress,
+                                               const uint8_t *     aBuffer,
+                                               uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+
+    using RouterAdv::Option;
+    using RouterAdv::PrefixInfoOption;
+    using RouterAdv::RouterAdvMessage;
+
+    bool           needReevaluate = false;
+    const uint8_t *optionsBegin;
+    uint16_t       optionsLength;
+    const Option * option;
+
+    VerifyOrExit(aBufferLength >= sizeof(RouterAdvMessage));
+
+    otLogInfoBr("received Router Advertisement from %s on interface %u", aSrcAddress.ToString().AsCString(),
+                mInfraIfIndex);
+
+    optionsBegin  = aBuffer + sizeof(RouterAdvMessage);
+    optionsLength = aBufferLength - sizeof(RouterAdvMessage);
+
+    option = nullptr;
+    while ((option = Option::GetNextOption(option, optionsBegin, optionsLength)) != nullptr)
+    {
+        const PrefixInfoOption *pio;
+        Ip6::Prefix             prefix;
+
+        if (option->GetType() != Option::Type::kPrefixInfo)
+        {
+            continue;
+        }
+
+        pio = static_cast<const PrefixInfoOption *>(option);
+        pio->GetPrefix(prefix);
+        if (!IsValidOnLinkPrefix(prefix))
+        {
+            otLogInfoBr("ignore invalid prefix in PIO: %s", prefix.ToString().AsCString());
+            continue;
+        }
+
+        if (pio->GetValidLifetime() == 0)
+        {
+            if (mDiscoveredOnLinkPrefix == prefix)
+            {
+                otLogInfoBr("invalidate discovered on-link prefix %s", prefix.ToString().AsCString());
+                mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+                mDiscoveredOnLinkPrefix.Clear();
+                needReevaluate = true;
+            }
+        }
+        else
+        {
+            otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
+                        prefix.ToString().AsCString(), pio->GetValidLifetime());
+
+            // We keep tracking the latest on-link prefix.
+            mDiscoveredOnLinkPrefixInvalidTimer.Start(GetPrefixExpireDelay(pio->GetValidLifetime()));
+            mDiscoveredOnLinkPrefix = prefix;
+
+            // Stop Router Solicitation if we found a valid on-link prefix.
+            // Otherwise, we wait till the Router Solicitation process timeouted.
+            // So the maximum delay before the Border Router starts advertising
+            // its own on-link prefix is 9 (4 + 4 + 1) seconds.
+            mRouterSolicitTimer.Stop();
+            needReevaluate = true;
+        }
+    }
+
+    if (needReevaluate)
+    {
+        EvaluateRoutingPolicy();
```

Comment:
We may also need to add delay before this `EvaluateRoutingPolicy` because `RA` is a global event for all BRs on the same infrastructure link.

Reply:
SGTM. But prefer adding it in separate PR.